kobotoolbox/kpi
# Record all logins as access logs

## Checklist

1. [x] If you've added code that should be tested, add tests
2. [ ] If you've changed APIs, update (or create!) the documentation
3. [x] Ensure the tests pass
4. [x] Make sure that your code lints and that you've followed [our coding style](https://github.com/kobotoolbox/kpi/blob/master/CONTRIBUTING.md)
5. [x] Write a title and, if necessary, a description of your work suitable for publishing in our [release notes](https://community.kobotoolbox.org/tag/release-notes)
6. [ ] Mention any related issues in this repository (as #ISSUE) and in other repositories (as kobotoolbox/other#ISSUE)
7. [ ] Open an issue in the [docs](https://github.com/kobotoolbox/docs/issues/new) if there are UI/UX changes

## Description

Create an access log every time someone logs in.

## Notes
Adds a method to the AuditLog class to create a new access log with details about the authentication of a request. It will store the usual AuditLog fields, plus metadata about the source of the request and, in the case of django-loginas, the super user who actually made the request (this is stored in request._cached_user). The method can also take a user, which is useful for rare cases when request.user is empty (like in tests). The last optional parameter is an authentication type, which can be customized to identify the authentication as Token or Digest or similar.

Also adds a receiver to the user_logged_in signal to call the new method. The signal automatically passes the request along so we don't have to worry about that. There are tests for a simple login flow, loginas, MFA, and SSO login. A future PR will handle creating access logs for one-time authentication like token, Oauth, digest, etc.

To test:
1. login using normal login, sso, or mfa
2. in a python shell or in postgres, check that a new audit log has been created of type 'auth'
3. using a superuser account, login as another user from django-admin
4. check that a new audit log has been created recording the login of the new user, but with the super user name  and uid in the metadata
